shark-android/src/test/java/shark/AndroidReferenceMatchersTest.kt

```diff
@@ -0,0 +1,147 @@
+    expect(HMD_GLOBAL, 29, false)
+  }
+
+  @Test fun `Expect NVIDIA unknown XIAMI__resources_impl`() {
+    expect(NVIDIA, 30, false)
+  }
+
+  @Test fun `Expect XIAOMI known XIAMI__resources_impl`() {
+    expect(XIAOMI, 30, true)
+  }
+
+  @Test fun `Expect LENOVO known XIAMI__resources_impl`() {
+    expect(LENOVO, 30, true)
+  }
+
+  @Test fun `Expect INFINIX known XIAMI__resources_impl`() {
+    expect(INFINIX, 30, true)
+  }
+
+  @Test fun `Expect HMD_GLOBAL known XIAMI__resources_impl`() {
+    expect(HMD_GLOBAL, 30, true)
+  }
+
+  /**
+   * Validate whether or not we see a known library leak for the given device manufacturer
+   * and sdkInt combination, resulting from a call to
+   * AndroidReferenceMatchers.XIAMI__RESOURCES_IMPL.add(matchers).
+   */
+  private fun expect(manufacturer: String, sdkInt: Int, expectKnown: Boolean) {
+
+    val hprofBytes = dumpToBytes {
+      clazz("android.os.Build",
+        staticFields = listOf("MANUFACTURER" to string(manufacturer), "ID" to string("someId"))
+      )
+      clazz("android.os.Build\$VERSION",
+        staticFields = listOf("SDK_INT" to IntHolder(sdkInt))
+      )
+      val leaking = instance(clazz("android.content.Context"))
+      keyedWeakReference(leaking)
+      clazz(
+        "android.content.res.ResourcesImpl",
+        staticFields = listOf(
+          "mAppContext" to leaking
+        )
+      )
+    }
+
+    val matchers = mutableListOf<ReferenceMatcher>()
+    AndroidReferenceMatchers.XIAMI__RESOURCES_IMPL.add(matchers)
+    val actualMatch = matchers.first()
+    val analysis = hprofBytes.checkForLeaks<HeapAnalysisSuccess>(referenceMatchers = matchers)
+
+    if(expectKnown) {
+      // Expect no application leaks found
+      assertThat(analysis.applicationLeaks)
+        .describedAs(analysis.applicationLeaks.toString()).isEmpty()
+
+      // Expect we see a known library leak instead matching the leak pattern we expect.
+      assertThat(analysis.libraryLeaks).hasSize(1)
+      val leak = analysis.libraryLeaks.first()
+      assertThat(leak.pattern).isEqualTo(actualMatch.pattern)
+    } else {
+      // Expect that we don't know about this leak
+      assertThat(analysis.libraryLeaks).isEmpty()
+
+      // And hence it shows up as an application leak.
+      assertThat(analysis.applicationLeaks)
+        .describedAs(analysis.applicationLeaks.toString()).hasSize(1)
+      val leak = analysis.applicationLeaks.first()
+
+      val (expectedClassName, expectedRefName) = leak.leakTraces.first().referencePath.first().let {
+        it.owningClassName to it.referenceName
+      }
+
+      val (actualClassName, actualRefName) = (actualMatch.pattern as StaticFieldPattern).let {
+        it.className to it.fieldName
+      }
+      assertThat(expectedClassName).isEqualTo(actualClassName)
+      assertThat(expectedRefName).isEqualTo(actualRefName)
+    }
+  }
+
+  fun <T : HeapAnalysis> ByteArray.checkForLeaks(
```

Comment:
Maybe this should go into a test helper or extensions class alongside of AndroidReferenceMatchersTest??  

Since this is the only place we're using this here in the module, I left it here for now and thought we could move it to a common place if/when we need to do so